src/globus_sdk/experimental/globus_app/validating_token_storage/validators.py

```diff
@@ -0,0 +1,289 @@
+        dependent scopes before storage.
+
+        :param token_data_by_resource_server: The data to validate.
+        :param context: The validation context object, containing state of the system at
+            the time of validation.
+        """
+        identity_id = context.token_data_identity_id or context.prior_identity_id
+        for resource_server, token_data in token_data_by_resource_server.items():
+            self._validate_token_data_meets_scope_requirements(
+                resource_server=resource_server,
+                token_data=token_data,
+                identity_id=identity_id,
+                eval_dependent=False,
+            )
+
+    def after_retrieve(
+        self,
+        token_data_by_resource_server: t.Mapping[str, TokenStorageData],
+        context: TokenValidationContext,
+    ) -> None:
+        """
+        Validate the token data against scope requirements, including dependent
+        scope requirements.
+
+        :param token_data_by_resource_server: The data to validate.
+        :param context: The validation context object, containing state of the system at
+            the time of validation.
+        """
+        identity_id = context.token_data_identity_id or context.prior_identity_id
+        for token_data in token_data_by_resource_server.values():
+            self._validate_token_data_meets_scope_requirements(
+                resource_server=token_data.resource_server,
+                token_data=token_data,
+                identity_id=identity_id,
+            )
+
+    def _validate_token_data_meets_scope_requirements(
+        self,
+        *,
+        resource_server: str,
+        token_data: TokenStorageData,
+        identity_id: str | None,
+        eval_dependent: bool = True,
+    ) -> None:
+        """
+        Given a particular resource server/token data, evaluate whether the token +
+            user's consent forest meet the attached scope requirements.
+
+        Note: If consent_client was omitted, only root scope requirements are validated.
+
+        :param resource_server: The resource server string to validate against.
+        :param token_data: The token data to validate against.
+        :param identity_id: The identity ID of the user, from the surrounding context.
+        :param eval_dependent: Whether to evaluate dependent scope requirements.
+        :raises UnmetScopeRequirements: if token/consent data does not meet the
+            attached root or dependent scope requirements for the resource server.
+        :returns: None if all scope requirements are met (or indeterminable).
+        """
+        required_scopes = self.scope_requirements.get(resource_server)
+
+        # Short circuit - No scope requirements are, by definition, met.
+        if required_scopes is None:
+            return
+
+        # 1. Does the token meet root scope requirements?
+        root_scopes = token_data.scope.split(" ")
```

Comment:
Under ideal circumstances, scopes should be whitespace normalized. Is there any way that the scope in `token_data` might be from an untrusted source that has two or more spaces next to each other?

```pycon
>>> "a  b".split(" ")
['a', '', 'b']

>>> "a  b".split()
['a', 'b']
```

```suggestion
        root_scopes = token_data.scope.split()
```

Reply:
It looks like a bug to me at first, but if these data are from Auth then there will never be a double space unless a user manipulates the data directly.

I think a bare split call is more robust and better, but it's also an unreachable condition under most realistic usage. So I intend to apply the change but not to try to define a break-fix test case. Please unresolve if you disagree and want a test for it.

Reply:
No need for a test for this, I just prefer bare `.split()` calls when the intent is to split on whitespace. When an argument is given, I assume it's a critical element.

Thanks for updating this!